addons/website_slides/models/slide_channel.py

```diff
@@ -748,6 +888,20 @@ def _action_request_access(self, partner):
                 )
         return activities
 
+    def _get_channel_invitation_link(self, partner_id):
+        ''' This will return the url used as hyperlink for the channel invitiation email in template
+        mail_notification_channel_invite :
+          * The invitation route if user exists and is active (see slide_channel_invite)
+          * A partner-token based password creation page if res.partner partner_id has no active user yet,
+            redirecting to slide_channel_invite. '''
+        if partner_id.user_ids.state == 'active':
+            return '%s/slides/%s/invite' % (self.get_base_url(), self.id)
+        else:
+            if partner_id.signup_prepare():
+                signup_url = partner_id._get_signup_url_for_action()[partner_id.id]
+                return '%s%s' % (signup_url, '&redirect=/slides/%s/invite' % self.id)
```

Comment:
If public visibility, then
- if the partner has a user, I redirect him to the page, adding 'auth_login' in the url, adding it in the session when clicked, so it can browse pages freely and when connecting it will be prefilled.
- If not, then I redirect him to the course page, adding 'auth_signup_token' in the url, adding it to the session when clicked, so it can browse pages but when asked to create an account, it will be prefilled and the link is made with the corresponding partner
  -  However _If no free sign up is allowed_, I kept what I did. I use the direct _get_signup_url_for_action() method to set an invitation link
  
If a course has member only visibility I kept what I did. I use the direct _get_signup_url_for_action() method to set an invitation link if partner has no user. Otherwise, I just add to 'auth_login' in url. The rest happens in /invite route. First it is redirected to login page, then back to course if the user has access rights (e.g. is a member) or to main course page otherwise.